### DISCUSSION TITLE
Immich running on Nomad orchestrator

### DISCUSSION THREAD

**Discussion by dani** [immich-app/immich#6247] [Show and tell]
If anyone is interested, I've made a Nomad job to run Immich (on a complete hashicorp stack, using Nomad for the orchestrator, Consul for the service discovery and mesh, Vault for secrets management). It's available here https://git.lapiole.org/nomad/immich/src/branch/master/example

**Comment by matthiasschoger** [comment]
To add to that, I'd like to share my Immich setup on the Hashistack
https://github.com/matthiasschoger/hashilab-apps/tree/master/immich

In addition to the features listed by dani, it also includes Prometheus monitoring and the option to scale the worker and ML jobs across nodes.

**Comment by gbolo** [comment]
nice work

**Reply by matthiasschoger** [reply]
Thanks! Did learn quite a bit about Nomad while doing this, for example the ephemeral disk feature which is perfect for caching stuff like the AI models.

**Comment by ppenguin** [comment]
@dani @matthiasschoger Thanks, I'm just looking into this and already learned a lot from a quick glance. (I have 3 nomad clusters deployed over last year and it was a huge learning curve, mainly because I didn't find much clear learning resources and nomad still seems quite niche. Additionally, I run nomad on top of NixOS on bare-metal and VPS, which was an added challenge.)
Sorry in advance for hijacking this, but would you have any pointers to consistent resources about "do's and don'ts / howtos" in nomad? Also:
- `proxy`: nice, I was always getting by with `DATABASE_URL={{ range services "svc-mypg" }}...`, but I suppose using `proxy` could be "best practice"?
- `consul` DNS never worked for me, any good resources?
- out of interest: what CSI backends are you using? I'm on seaweedfs myself, but I don't yet trust it enough for important data. It's also ballooning pretty severely.

Cheers

**Reply by matthiasschoger** [reply]
Hi @ppenguin ,

yeah, Nomad can be quite a journey. I really like the simplicity compared to k8s, but the online resources available just handle very simple use-cases. A reference HA setup would be great.

I tried to do something like that with my "core" repo at 
https://github.com/matthiasschoger/hashilab-core
Not sure if it's really best practice, but I think it works great for me and my use-case.

- proxy: I think the overlay network can be considered "best practice" for Nomad + Consul. It's much more flexible since it does not require a restart of Immich if, for example, the database or Redis move to a different machine.
- consul: Regarding DNS, have a look at my CoreDNS job in hashilab-core. It's my main DNS and forwards all .consul DNS request to the local Consul instance.
- CSI: Just the vanilla NFS from registry.k8s.io/sig-storage/nfsplugin:v4.9.0. I'm sure there a better plugins out there, but this works one for me.

Regarding howtos, I'm afraid I can help with that. Read a ton of article on the internet, but nothing which I would say is essential. Lately I started to use ChatGPT, which is surprisingly good.
Played around with NixOS myself, but could not really get it to work with the Nomad CNI drivers and fell back to Debian Trixie. Care to share your NixOS setup?

**Reply by ppenguin** [reply]
> Played around with NixOS myself, but could not really get it to work with the Nomad CNI drivers and fell back to Debian Trixie. Care to share your NixOS setup?

The full multi-server flake is not ready for publication, but I pasted the one-node (homelab) `nomad` config in [this gist](https://gist.github.com/ppenguin/5a85f502b62f95dfd9d981484990d34e), which is pretty much self-contained. HTH.

The rationale of the NixOS setup is to have infrastructure (i.e. the nomad/consul/vault services themselves) under NixOS, and everything that runs on it as normal nomad jobs. It has been working pretty well, even in a somewhat more complicated 4-VPS cluster setup (3 clients + 1 server) connected by wiregard for the internal nomad/consul network. (That's also where NixOS shines, because the client nodes are pretty much identical so provisioning is a breeze and pretty much 0-config).

About two months ago I had a regression on `nixos-unstable` where one CNI-driver suddenly stopped working, so I quickly switched to `nixos-25.05` for the time being, which worked. Haven't checked back since then.

**Reply by matthiasschoger** [reply]
Just checked the gist, thanks for posting it.

Looks like I will need to dig quite a bit deeper into Nix to understand fully what you did there. 
When I was playing around, I tried to pull everything from stable, but only Nomad, Consul and CNI from unstable. Maybe your approach is the better way.

Will probably try again after my migration to Trixie.
Would be great if you could ping me when you think your full flake is ready for publication.

**Reply by ppenguin** [reply]
@matthiasschoger Even though mixed inputs (unstable/stable) is theoretically perfectly fine in NixOS, it has given me more trouble than benefit in the past (therefore I now simply equate `npkgs = pkgs`). For the same reason it's also advisable to use overlays sparingly or at least with their implications in mind (mainly the fact that they can cause a chain-reaction of source-built packages if you change by overlay a package that many others depend on).

As for the complexity: the main somewhat more elaborate tricks I used are only to make provisioning fully declerative by having all (or most -> WIP) certificate files included in the flake; the rest of the config is pretty standard. 

I can ping if published full config, but I'm afraid it's not a priority right now. Even though I have all my secrets in sops-nix, I'm as a matter of principle a bit uncomfortable to make infrastructure (i.e. hosts, names, network structure) publicly accessible, so it would require split flakes to realize this in line with security expectations. This is on the wish list but I won't find time for that soon.

**Reply by dani** [reply]
> @dani @matthiasschoger Thanks, I'm just looking into this and already learned a lot from a quick glance. (I have 3 nomad clusters deployed over last year and it was a huge learning curve, mainly because I didn't find much clear learning resources and nomad still seems quite niche.

Indeed, very few resources (compared to eg, K8S). You can take a look at my repos here https://git.lapiole.org/nomad/. In each repo, you'll find an example dir which has a full job almost ready to use (with all the needed vault policies, roles, consul policies etc.)

> Additionally, I run nomad on top of NixOS on bare-metal and VPS, which was an added challenge.) Sorry in advance for hijacking this, but would you have any pointers to consistent resources about "do's and don'ts / howtos" in nomad? Also:
> 
> * `proxy`: nice, I was always getting by with `DATABASE_URL={{ range services "svc-mypg" }}...`, but I suppose using `proxy` could be "best practice"?

Yes, using the Consul Services Mesh is best practice. 

> * `consul` DNS never worked for me, any good resources?

In my case, all my hosts are using an unbound instance as resolver, and this unbound instance forward the .consul domain to my 3 consul servers

> * out of interest: what CSI backends are you using? I'm on seaweedfs myself, but I don't yet trust it enough for important data. It's also ballooning pretty severely.

I'm using democratic-csi (with local, freenas-api NFS and iSCSI drivers), and CephCSI (both RBD and CephFS). Also played with RocketDuck for NFS which was also working, and quite trivial to setup

**Reply by ppenguin** [reply]
@dani Thanks for your elaboration. I noticed that your `.nomad.hcl` files seem to be somehow additionally templated (`[[ ... ]]`), how/by which tooling/template engine is this used?

**Reply by dani** [reply]
Yes. I'm using a custom pre-processing engine (based on gomplate) because nomad-pack is not actively developped, and far from enough for my needs. The tool itself is open, but I haven't written any documentation for it yet. Anyway, here it is : https://git.lapiole.org/dani/ctctl

It's working by relying on re-usable templates which can be imported (via git submodules), handles project dependencies, multi-level YAML config to adapt to the target env, switching between Nomad clusters etc.
